tools/config-gen/main.go

```diff
@@ -40,34 +41,37 @@ type templateData struct {
 
 func validateFlags() error {
 	if *paramsFile == "" {
-		return fmt.Errorf("input filename cannot be empty")
+		return fmt.Errorf("params filename cannot be empty")
 	}
-	if *outFile == "" {
-		return fmt.Errorf("output filename cannot be empty")
+	if *outDir == "" {
+		return fmt.Errorf("output directory cannot be empty")
 	}
-	if *templateFile == "" {
-		return fmt.Errorf("template filename cannot be empty")
+	if *templateDir == "" {
+		return fmt.Errorf("template directory cannot be empty")
 	}
 	return nil
 }
 
-func write(data templateData) error {
-	outputFile, err := os.Create(*outFile)
-	if err != nil {
-		defer outputFile.Close()
-	}
-	if err != nil {
-		return err
-	}
-	tmpl, err := template.New("config.tpl").ParseFiles(*templateFile)
-	if err != nil {
+// write applies the dataObj which could be an object of any type, to the
+// templateFile and writes the generated text into the outputFile.
+func write(dataObj any, outputFile, templateFile string) (err error) {
+	var outF *os.File
+	if outF, err = os.Create(outputFile); err != nil {
 		return err
 	}
-	err = tmpl.Execute(outputFile, data)
-	if err != nil {
+	defer func() {
+		closeErr := outF.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	file := path.Base(templateFile)
+	if tmpl, err := template.New(file).ParseFiles(templateFile); err == nil {
+		return tmpl.Execute(outF, dataObj)
+	} else {
```

Comment:
Should just move the `return err` out of the else-block and remove the else-block (as it's unnecessary as the if-block always returns anyway).

Reply:
I think the "err" within the "if" block shadows the "err" in the outer scope. So, modifications to the "err" within the if-block will probably have no effect on the "err" in the outer scope. That said, my understanding of Go-scopes could be wrong here. So, let me know if you think differently.